### PR TITLE
[SPARK-47581][CORE] SQL catalyst: Migrate logWarning with variables to structured logging framework

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Migrate logWarning with variables of the Catalyst module to structured logging framework. This transforms the logWarning entries of the following API
```
def logWarning(msg: => String): Unit
```
to
```
def logWarning(entry: LogEntry): Unit
```

### Why are the changes needed?

To enhance Apache Spark's logging system by implementing structured logging.

### Does this PR introduce _any_ user-facing change?

Yes, Spark core logs will contain additional MDC

### How was this patch tested?

Compiler and scala style checks, as well as code review.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by dtenedor** [comment]
@gengliangwang here is the structured logging migration for the `logWarning` calls within Catalyst per https://issues.apache.org/jira/browse/SPARK-47581.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala [hunk 1]
```diff
@@ -284,7 +287,8 @@ object StreamingJoinHelper extends PredicateHelper with Logging {
           Seq(negateIfNeeded(castedLit, negate))
         case a @ _ =>
           logWarning(
-            s"Failed to extract state value watermark from condition $exprToCollectFrom due to $a")
+            log"Failed to extract state value watermark from condition " +
+              log"${MDC(JOIN_CONDITION, exprToCollectFrom)} due to ${MDC(JOIN_CONDITION, a)}")
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala:291]: What is `a` here? It seems that you are using a same key for two variables

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVHeaderChecker.scala [hunk 2]
```diff
@@ -75,19 +75,21 @@ class CSVHeaderChecker(
           }
           if (nameInHeader != nameInSchema) {
             errorMessage = Some(
-              s"""|CSV header does not conform to the schema.
-                  | Header: ${columnNames.mkString(", ")}
-                  | Schema: ${fieldNames.mkString(", ")}
-                  |Expected: ${fieldNames(i)} but found: ${columnNames(i)}
-                  |$source""".stripMargin)
+              log"""|CSV header does not conform to the schema.
+                    | Header: ${MDC(COLUMN_NAME, columnNames.mkString(", "))}
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVHeaderChecker.scala:79]: We can't use COLUMN_NAME for 4 different variables...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 3]
```diff
@@ -74,10 +75,10 @@ object CharVarcharUtils extends Logging with SparkCharVarcharUtils {
     if (SQLConf.get.charVarcharAsString) {
       replaceCharVarcharWithString(dt)
     } else if (hasCharVarchar(dt)) {
-      logWarning("The Spark cast operator does not support char/varchar type and simply treats" +
-        " them as string type. Please use string type directly to avoid confusion. Otherwise," +
-        s" you can set ${SQLConf.LEGACY_CHAR_VARCHAR_AS_STRING.key} to true, so that Spark treat" +
-        s" them as string type as same as Spark 3.0 and earlier")
+      logWarning(log"The Spark cast operator does not support char/varchar type and simply treats" +
+        log" them as string type. Please use string type directly to avoid confusion. Otherwise," +
+        log" you can set ${MDC(SQL_CONF_KEY, SQLConf.LEGACY_CHAR_VARCHAR_AS_STRING.key)} " +
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:80]:
```suggestion
        log" you can set ${MDC(CONFIG, SQLConf.LEGACY_CHAR_VARCHAR_AS_STRING.key)} " +
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala [hunk 4]
```diff
@@ -321,8 +322,11 @@ object ResolveDefaultColumns extends QueryErrorsBase
           Option(casted.eval(EmptyRow)).map(Literal(_, targetType))
         } catch {
           case e @ ( _: SparkThrowable | _: RuntimeException) =>
-            logWarning(s"Failed to cast default value '$l' for column $colName from " +
-              s"${l.dataType} to $targetType due to ${e.getMessage}")
+            logWarning(log"Failed to cast default value '${MDC(COLUMN_DEFAULT_VALUE, l)}' " +
+              log"for column ${MDC(COLUMN_NAME, colName)} " +
+              log"from ${MDC(COLUMN_DATA_TYPE, l.dataType)} " +
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:327]: we can't use COLUMN_DATA_TYPE for two variables here.

**Review by gengliangwang — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala]: Sorry about this, I did not know we had to have unique keys for the log keys. Fixed now.

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 5]
```diff
@@ -95,6 +117,8 @@ object LogKey extends Enumeration {
   val SHUFFLE_MERGE_ID = Value
   val SIZE = Value
   val SLEEP_TIME = Value
+  val SLEEP_TIME_SECONDS = Value
```
Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:120]: This is not used.

Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:120]:
```suggestion
```

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by panbingkun** [comment]
Unfortunately, GA's master workflow has failed.
https://github.com/apache/spark/actions/runs/8611002654/job/23597440628

**Comment by panbingkun** [comment]
I submitted a follow-up PR, let it recover first
https://github.com/apache/spark/pull/45958

**Comment by dtenedor** [comment]
Thanks @panbingkun for the fix!!

**Pull request closed without merging**